design/deletion-plugins.md

```diff
@@ -0,0 +1,77 @@
+# Deletion Plugins
+
+
+## Abstract
+Velero should introduce a new type of plugin that runs when a backup is deleted.
+These plugins will delete any external resources associated with the backup so that they will not be left orphaned.
+
+## Background
+With the CSI plugin, Velero developers introduced a pattern of using BackupItemAction and RestoreItemAction plugins tied to PersistentVolumeClaims to create other resources to complete a backup.
+In the CSI plugin case, Velero does clean up of these other resources, which are Kubernetes Custom Resources, within the core Velero server.
+However, for external plugins that wish to use this same pattern, this is not a practical solution.
+Velero's core cannot be extended for all possible Custom Resources, and not external resources that get created are Kubernetes Custom Resources.
+
+Therefore, Velero needs some mechanism that allows plugin authors who have created resources within a BackupItemAction or RestoreItemAction plugin to ensure those resources are deleted, regardless of what system those resources reside in.
+
+## Goals
+- Provide a new plugin type in Velero that is invoked when a backup is deleted.
+
+## Non Goals
+- Implementations of specific deletion plugins.
+- Rollback of deletion plugin execution.
+
+
+## High-Level Design
+Velero will provide a new plugin type that is similar to its existing plugin architecture.
+These plugins will be referred to as `DeletionAction` plugins.
+`DeletionAction` plugins will receive the `Backup` CustomResource being deleted on execution.
+
+`DeletionAction` plugins cannot prevent deletion of an item.
+This is because multiple `DeletionAction` plugins can be registered, and this proposal does not include rollback and undoing of a deletion action.
+Thus, if multiple `DeletionAction` plugins have already run but another would request the deletion of a backup stopped, the backup that's retained would be inconsistent.
+
+`DeletionActions` will apply to `Backup`s based on a label on the `Backup` itself.
+In order to ensure that `Backup`s don't execute `DeletionAction` plugins that are not relevant to them, `DeletionAction` plugins can register an `AppliesTo` function which will define a label selector on Velero backups.
+
+## Detailed Design
```

Comment:
We should also mention that there will not be any ordering guarantees for the `DeleteAction` plugins when running multiple delete action plugins for a backup.

Reply:
We could make them alphanumerical by name, just so we have some known order.